src/backend/distributed/catalog/dependency.c

```diff
@@ -249,7 +287,42 @@ recurse_pg_depend(const ObjectAddress *target,
 		 * recurse depth first, this makes sure we call apply for the deepest dependency
 		 * first.
 		 */
-		recurse_pg_depend(&address, follow, apply, context);
+		recurse_pg_depend(&address, follow, expand, apply, context);
+
+		/*
+		 * the expand function will be used (when set) to add more candidates to the
+		 * exploration.
+		 */
+		if (expand != NULL)
+		{
+			List *expansion = expand(context, pg_depend);
+			ListCell *expandedObjectAddress = NULL;
+			foreach(expandedObjectAddress, expansion)
+			{
+				ObjectAddress *expandedAddress = (ObjectAddress *) lfirst(
+					expandedObjectAddress);
+				recurse_pg_depend(expandedAddress, follow, expand, apply, context);
+
+				if (apply != NULL)
+				{
+					/*
```

Comment:
Can we move this logic into a function just for readability purpose? I want `recurse_pg_depend ` be as simple as the initial commit :)

Reply:
I have refactored `expand` a bit to not return an `ObjectAddres *` but a `Form_pg_depend`, as if you are adding records to `pg_depend` based on the rules. This made this body easier to digest.

Do you think this maakes it easy enough again?